### PR TITLE
Added Euler problem 31 test

### PR DESCRIPTION
**Pull request by ashrko619**

### CODE REVIEW

**Comment by jorander** [comment]
Hi @ashrko619 ,
I think it is easier for @danieldietrich to merge your PR if you do one PR for each test instead of pushing them all together.
BR,
Jörgen

File: javaslang/src/test/java/javaslang/collection/euler/Euler31Test.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    public void shouldSolveProblem31() {
+        List<Integer> coins = List.of(1, 2, 5, 10, 20, 50, 100, 200);
+        assertThat(coinSums(200, coins)).isEqualTo(73682);
+    }
+
+    private static int coinSums(int n, List<Integer> coins) {
+        if (n == 0) return 1;
+        if (n < 0 || coins.isEmpty()) return 0;
+        return coinSums(n, coins.tail()) + coinSums(n - coins.head(), coins);
+    }
```
Comment (jorander) [inline javaslang/src/test/java/javaslang/collection/euler/Euler31Test.java]: Nice and simple solution to this problem!
However, I think we could show-case more of Javaslang's functionality by replacing the if:s with Pattern Matching. I also suggest using a memoized coinSum-function to enhance performance.

```java
private static int coinSums(int targetSum, List<Integer> coins) {
        return Match(Tuple.of(targetSum, coins)).of(
                Case(Tuple2($(0),$()), 1),
                Case(Tuple2($(ts -> ts < 0), $()), 0),
                Case(Tuple2($(), $(cx -> cx.isEmpty())), 0),
                Case($(), t -> memoizedCoinSums.apply(t._1, t._2.tail()) + memoizedCoinSums.apply(t._1 - t._2.head(), t._2))
        );
    }

private static final Function2<Integer, List<Integer>, Integer> memoizedCoinSums = Function2.of(Euler31Test::coinSums).memoized();
```

**Comment by danieldietrich** [comment]
@ashrko619 Thank you!

Indeed, splitting PRs makes life easier looking at features merged to src/main/java. Otherwise we have a problem when cherry picking features into specific releases. Also code review is simpler when having small PRs.

Could you please resolve the merge conflicts? I think I've pulled in overlapping code from another PR...

@ashrko619 @jorander I think we do not need to showcase more features in this special case. The code is already really concise - looks a bit like haskell :-D

**Comment by ashrko619** [comment]
I have been trying to push them as separate PRs but somehow I mange to screw up - git noob here :laughing: 
@danieldietrich sure will merge the conflicts :)

**Comment by danieldietrich** [comment]
@paplorinc do you have an idea how to reduce this PR to the relevant project euler 31 test? At the moment there are 21 files changed. I think something went wrong with the merge!? Could you provide @ashrko619 with a magic git command?

**Comment by paplorinc** [comment]
Hello @ashrko619, @danieldietrich :)

Sure, the simplest would probably be to pull the project from upstream and soft rebase it on head, making all the changes merged and not committed and can me committed in a single swoop :)
This can also be done from `Idea` by making it up-to-date first: `VCS` -> `git` -> `Rebase my GitHub fork`, solving the conflicts (it seems there aren't any) if there are any and from the `Version Control`s `Log` view do a right click and reset/soft to `upstream/master` and commit the modified files again :)

Hope that helps :)

**Comment by danieldietrich** [comment]
@ashrko619 many thanks for the PR! When looking at the changed files (21!) I have no good feeling merging it. Please copy/paste the relevant parts (Euler 31 Test) into a new PR. I will close this one.
I hope you understand.
Greets
Daniel

**Pull request closed without merging**